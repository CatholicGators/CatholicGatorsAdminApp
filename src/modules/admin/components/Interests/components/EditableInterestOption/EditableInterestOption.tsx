import React, { Component } from 'react'
import {
    Checkbox,
    Theme,
    createStyles,
    withStyles,
    IconButton,
    Input,
    Typography
} from '@material-ui/core'
import {
    Delete,
    Edit,
    Save,
    Close
} from '@material-ui/icons'
import { Option } from '../../Interests'

const styles = (theme: Theme) => createStyles({
    checkbox: {
        margin: 0,
        pointerEvents: 'none'
    },
    input: {
        margin: 0
    },
    optionRow: {
        display: 'flex',
        alignItems: 'center',
        '&:hover': {
            '& section': {
                display: 'block'
            }
        }
    },
    actions: {
        display: 'none',
        marginLeft: `${theme.spacing.unit}px`
    },
    editActions: {
        display: 'flex',
        alignItems: 'center'
    }
})

type Props = {
    isEditing: boolean,
    classes: any,
    option: Option,
    editOption: (optionId: number) => void,
    deleteOption: (optionId: number) => void,
    onSave: (optionId: number, newText: string) => void,
    onCancel: (optionId: number, newText: string) => void
}

type State = {
    editedOptionText: string
}

export class EditableInterestOption extends Component<Props, State> {
    state = {
        editedOptionText: this.props.option.text
    }

    onEditOptionChange(newText: string) {
        this.setState({ editedOptionText: newText })
    }

    render() {
        const {
            isEditing,
            classes,
            option,
            editOption,
            deleteOption,
            onSave,
            onCancel
        } = this.props

        const {
            editedOptionText
        } = this.state

        return (
            <div className={classes ? classes.optionRow : null}>
                <Checkbox
                    className={classes ? classes.checkbox : null}
                    color='primary'
                />
                {isEditing ? ([
                    <Input
                        fullWidth
                        key={1}
                        className={classes ? classes.input : null}
                        value={this.state.editedOptionText}
                        onChange={ev => this.onEditOptionChange(ev.target.value)}
                    />,
                    <div key={2} className={classes ? classes.editActions : null}>
                        <IconButton onClick={() => onSave(option.id, editedOptionText)}>
                            <Save />
                        </IconButton>
                        <IconButton onClick={() => onCancel(option.id, editedOptionText)}>
                            <Close />
                        </IconButton>
                    </div>
                ]) : ([
                    <Typography key={1}>{option.text}</Typography>,
                    <section key={2} className={classes ? classes.actions : null}>
                        <IconButton onClick={() => editOption(option.id)}>
                            <Edit />
                        </IconButton>
                        <IconButton onClick={() => deleteOption(option.id)}>
                            <Delete />
                        </IconButton>
                    </section>
                ])}
            </div>
        )
    }
}

export default withStyles(styles)(EditableInterestOption)
