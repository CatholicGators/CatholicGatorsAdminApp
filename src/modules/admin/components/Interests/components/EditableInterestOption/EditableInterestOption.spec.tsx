import React from 'react'
import { shallow } from 'enzyme';

import { EditableInterestOption, Props } from './EditableInterestOption'
import {
    Edit,
    Delete,
    Save,
    Close
} from '@material-ui/icons';
import { Input } from '@material-ui/core';

describe('EditableInterestOption', () => {
    let props: Props, wrapper, instance: EditableInterestOption

    beforeEach(() => {
        props = {
            classes: {},
            isEditing: false,
            option: {
                id: 1,
                text: "totally testing"
            },
            beginEditing: jest.fn(),
            cancelEditing: jest.fn(),
            deleteOption: jest.fn(),
            saveOption: jest.fn()
        }
        wrapper = shallow(<EditableInterestOption {...props} />)
        instance = wrapper.instance()
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('defaults the editedOptionText to the text of the option that is passed in', () => {
        expect(wrapper.state().editedOptionText).toBe(props.option.text)
    })

    describe('when not editing', () => {
        it('renders the edit and delete buttons', () => {
            expect(wrapper.exists(Edit)).toBe(true)
            expect(wrapper.exists(Delete)).toBe(true)
            expect(wrapper.exists(Save)).toBe(false)
            expect(wrapper.exists(Close)).toBe(false)
            expect(wrapper.exists(Input)).toBe(false)
        })

        it('beginEdit is called when edit clicked', () => {
            const edit = wrapper.find('#edit')
            edit.simulate('click')

            expect(props.beginEditing).toHaveBeenCalledWith(props.option.id)
        })

        it('cancelEdit is called when cancel clicked', () => {
            const deleteButton = wrapper.find('#delete')
            deleteButton.simulate('click')

            expect(props.deleteOption).toHaveBeenCalledWith(props.option.id)
        })
    })

    describe('when editing', () => {
        beforeEach(() => {
            wrapper.setProps({ isEditing: true })
        })

        it('renders the input, save, and cancel buttons', () => {
            expect(wrapper.exists(Edit)).toBe(false)
            expect(wrapper.exists(Delete)).toBe(false)
            expect(wrapper.exists(Save)).toBe(true)
            expect(wrapper.exists(Close)).toBe(true)
            expect(wrapper.exists(Input)).toBe(true)
        })

        it('saveOption is called when save clicked', () => {
            const newText = 'test'
            const input = wrapper.find(Input)
            input.simulate("change", { target: { value: newText } });
            const save = wrapper.find('#save')
            save.simulate('click')

            expect(props.saveOption).toHaveBeenCalledWith(props.option.id, newText)
        })

        it('cancelEdit is called when cancel clicked', () => {
            const cancel = wrapper.find('#cancel')
            cancel.simulate('click')

            expect(props.cancelEditing).toHaveBeenCalled()
        })
    })
})
