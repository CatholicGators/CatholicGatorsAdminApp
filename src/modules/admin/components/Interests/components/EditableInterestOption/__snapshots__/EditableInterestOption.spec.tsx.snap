// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableInterestOption should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableInterestOption
    beginEditing={[MockFunction]}
    cancelEditing={[MockFunction]}
    classes={Object {}}
    deleteOption={[MockFunction]}
    isEditing={false}
    option={
      Object {
        "id": 1,
        "text": "totally testing",
      }
    }
    saveOption={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(Checkbox)
          color="primary"
        />,
        Array [
          <WithStyles(Typography)>
            totally testing
          </WithStyles(Typography)>,
          <section>
            <WithStyles(IconButton)
              id="edit"
              onClick={[Function]}
            >
              <pure(EditIcon) />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              id="delete"
              onClick={[Function]}
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>
          </section>,
        ],
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": undefined,
          "color": "primary",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "children": "totally testing",
        },
        "ref": null,
        "rendered": "totally testing",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithStyles(IconButton)
              id="edit"
              onClick={[Function]}
            >
              <pure(EditIcon) />
            </WithStyles(IconButton)>,
            <WithStyles(IconButton)
              id="delete"
              onClick={[Function]}
            >
              <pure(DeleteIcon) />
            </WithStyles(IconButton)>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <pure(EditIcon) />,
              "id": "edit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <pure(DeleteIcon) />,
              "id": "delete",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "section",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Checkbox)
            color="primary"
          />,
          Array [
            <WithStyles(Typography)>
              totally testing
            </WithStyles(Typography)>,
            <section>
              <WithStyles(IconButton)
                id="edit"
                onClick={[Function]}
              >
                <pure(EditIcon) />
              </WithStyles(IconButton)>
              <WithStyles(IconButton)
                id="delete"
                onClick={[Function]}
              >
                <pure(DeleteIcon) />
              </WithStyles(IconButton)>
            </section>,
          ],
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": undefined,
            "color": "primary",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "children": "totally testing",
          },
          "ref": null,
          "rendered": "totally testing",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(IconButton)
                id="edit"
                onClick={[Function]}
              >
                <pure(EditIcon) />
              </WithStyles(IconButton)>,
              <WithStyles(IconButton)
                id="delete"
                onClick={[Function]}
              >
                <pure(DeleteIcon) />
              </WithStyles(IconButton)>,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <pure(EditIcon) />,
                "id": "edit",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <pure(DeleteIcon) />,
                "id": "delete",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "section",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
